Add a blog post about secure MCP SSE server

sberyozkin opened the pull request.

cescoffier commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,202 @@
+
+== Introduction
+
+https://modelcontextprotocol.io/specification/2025-03-26[The latest version of the Model Context Protocol (MCP) specification] introduces an https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization[authorization] flow for HTTP based MCP transports.
```
Are you sure it can be applied to the legacy HTTP transport?

cescoffier commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,202 @@
+
+https://modelcontextprotocol.io/specification/2025-03-26[The latest version of the Model Context Protocol (MCP) specification] introduces an https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization[authorization] flow for HTTP based MCP transports.
+
+MCP authorization will enable MCP SSE and Streamable HTTP clients to access MCP servers securely and is being very actively discussed and evaluated right now. This https://github.com/modelcontextprotocol/modelcontextprotocol/pull/284[MCP authorization PR] is one of the focal points for the MCP authorization discussions.
```
I dont understand the first sentence. First, I would remove securely - as it is only about authentication.

cescoffier commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,202 @@
+
+MCP authorization will enable MCP SSE and Streamable HTTP clients to access MCP servers securely and is being very actively discussed and evaluated right now. This https://github.com/modelcontextprotocol/modelcontextprotocol/pull/284[MCP authorization PR] is one of the focal points for the MCP authorization discussions.
+
+While it will take a bit of time for the MCP authorization specificaion to finalize and be widely supported, one thing you can be certain about is that secure MCP servers that wish to interoperate with MCP authorization aware MCP clients will be required to accept and verify bearer access tokens sent by these clients on behalf of their users.
```
Here, there is an issue - so far (except if I missed it), this is not part of the capabilities of the server. Thus, the client must now that the server requires authentication. Also, I'm not sure about multi-users, in the sense that the MCP client may have a "connection" (which is actually a client in the spec) with a server. It's unclear if these can be used by multiple users, or if you need to instantiate one "client" per user.

cescoffier commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,202 @@
+
+While it will take a bit of time for the MCP authorization specificaion to finalize and be widely supported, one thing you can be certain about is that secure MCP servers that wish to interoperate with MCP authorization aware MCP clients will be required to accept and verify bearer access tokens sent by these clients on behalf of their users.
+
+The https://github.com/quarkiverse/quarkus-mcp-server[Quarkus MCP Server project] is ready for you to start experimenting with secure MCP SSE servers.
```
We would need a compatible client to build a story.

maxandersen replied:
just to clarify - I thought quarkus-mcp-server did *not* yet implement 2025-03-26 spec.

are we just saying here that the oauth in the new spec is possible because it literally just is oauth which is flexible enough in quarkus to apply here or?

maxandersen replied:
now i reread it - this is explaining how to secure mcp servers in general - and it does NOT require the latest version of the spec nor a compatible client (besides it should be able to send a header token).

I think that should be made clearer in that introduction that this is something that works for MCP *today*.

sberyozkin replied:
@cescoffier @maxandersen Thanks, yes, this post is about showing how, given a token, users can access Quarkus MCP server from either Dev UI or curl, given that none of the existing MCP clients can actually use OAuth2 to authenticate users and forward tokens automatically.

>I think that should be made clearer in that introduction that this is something that works for MCP today.

I'm not sure about it, I believe we don't have MCP clients that can automatically consume MCP SSE server which responds 401 if no token is sent

sberyozkin replied:
@maxandersen Sure, this is just not the MCP authorization spec but as agreed we will cover it

cescoffier commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,202 @@
+----
+
+[[mcp-server-devui]]
+=== Access MCP SSE server in DevUI
```
```suggestion
=== Access MCP SSE server in Dev UI
```

cescoffier commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,202 @@
+
+Select an OpenId Connect card and https://quarkus.io/guides/security-openid-connect-dev-services#develop-service-applications[login to Keycloak] using an `alice` name and an `alice` password.
+
+What about other providers such as `Auth0` or https://quarkus.io/guides/security-openid-connect-providers#github[GitHub] ?
```
That should be an admonition.

cescoffier commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,202 @@
+----
+<1> Enforce an authenticated access to the main MCP SSE endpoint. See also how the tool is secured with an annotation in the <<tool>> section above, though you can also secure access to the tool by also listing an SSE tools endpoint in the configuration, for example: `quarkus.http.auth.permission.authenticated.paths=/mcp/sse,/mcp/messages/*`.
+
+What about the OIDC configuration ? It is provided for you in devmode by https://quarkus.io/guides/security-openid-connect-dev-services[Dev Services for Keycloak].  It creates a default realm, client and adds two users, `alice` and `bob`, for you to get started immediately. You can also register a custom Keycloak realm.
```
That should be an admonition.

cescoffier commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,202 @@
+Select an OpenId Connect card and https://quarkus.io/guides/security-openid-connect-dev-services#develop-service-applications[login to Keycloak] using an `alice` name and an `alice` password.
+
+What about other providers such as `Auth0` or https://quarkus.io/guides/security-openid-connect-providers#github[GitHub] ?
+No problems, you can login to them from OIDC DevUI as well, for example, see how you can https://quarkus.io/guides/security-oidc-auth0-tutorial#looking-at-auth0-tokens-in-the-oidc-dev-ui[login to Auth0] - the only requirement is to update your provider application registration to allow callbacks to DevUI.
```
```suggestion
No problems, you can login to them from OIDC Dev UI as well, for example, see how you can https://quarkus.io/guides/security-oidc-auth0-tutorial#looking-at-auth0-tokens-in-the-oidc-dev-ui[login to Auth0] - the only requirement is to update your provider application registration to allow callbacks to DevUI.
```

cescoffier commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,202 @@
+
+Make a tool call by selecting the `Call` action and get the response:
+
+image::mcp_server_tool_response.png[MCP Server tool response,align="center"]
```
I think you need to explain in which message the token is passed - it is during the handshake, during the tool invocation...

maxandersen replied:
since this is for MCP in general not the new spec - the token is set on all the headers afaik.

but yes worth highlighting.

sberyozkin replied:
@maxandersen @cescoffier 
I've tried to highlight it when descriving how to create MCP server, as otherwise we'd need to repeat it in all 3 sections, `curl`, `MCP inspector` and Dev UI

cescoffier commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,202 @@
+
+image::mcp_server_tool_response.png[MCP Server tool response,align="center"]
+
+=== Access MCP SSE server with curl
```
Actually, here you explain it a bit better - but you need to introduce where is happens in the protocol.

cescoffier commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,202 @@
+
+[source,shell]
+----
+url -v -H "Authorization: Bearer ey..." -H "Content-Type: application/json" --data @call.json http://localhost:8080/mcp/messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
```
```suggestion
curl -v -H "Authorization: Bearer ey..." -H "Content-Type: application/json" --data @call.json http://localhost:8080/mcp/messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
```

cescoffier commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,202 @@
+event: message
+data: {"jsonrpc":"2.0","id":2,"result":{"isError":false,"content":[{"text":"alice","type":"text"}]}}
+----
+
```
This is the "SSE transport" - it would be interesting to see how it works with the streamable HTTP.

Note that your "flow" here, does not do the handshake, so it not very "compliant" - it works because we had to relax that constraint.

maxandersen replied:
no streamable http in old version.

sberyozkin replied:
MCP server does not currently support it as well, Martin had some concerns

cescoffier reviewed (changes requested):
It's a great post. I thing we can clarify a few things:

1. Explains where the authentication happens in the MCP protocol (in which message the header must be set), even diving a bit in the MCP protocol a bit
2. I would actually start with curl and then having a dev ui demo (maybe even a quick video)

maxandersen commented on _posts/2025-04-23-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,205 @@
+data: {"jsonrpc":"2.0","id":2,"result":{"isError":false,"content":[{"text":"alice","type":"text"}]}}
+----
+
+== Conclusion
```
haven't tried - but one curious thing to maybe highlight is what happens if you run this server in stdio mode - is the securecontext ignored or will it make it fail? (I assume/hope the latter and have some way to make it ignored)

sberyozkin replied:
@maxandersen for a tool like this one to work with Stdio the MCP Stdio transport would have to support carrying the token as part of the call data I guess. If the tool must be secured then it should be secured no matter what transport is used to access it

sberyozkin commented:
@maxandersen @cescoffier Thanks very much for the thoughtful review, I know I was not quite getting along, but as always, on reflection, I know the updated PR should be better. 

One thing I found difficult to update was to move `curl` as the first example. MCP Dev UI is good as the first one because both MCP and OIDC cards are aligned and I describe how to get OIDC token, and use Dev UI, and then in the follow up section I just refer to the Dev UI section for users to see how to get the token. Starting from curl or the inspector requires to repeat the same info re the OIDC token acqusitiuon.

I honestly don't mind if it goes first, but not sure how to make it work neatly if curl goes first, Clement, if you'd like please give it a try.

Please follow up tomorrow morning with suggestions, especially with respect to some further technical clarifications and we should be ready to merge soon.

Cheers

sberyozkin commented:
I've resolved all the comments though I may not have addressed them as you were expecting - please create new suggestions

sberyozkin commented:
May be I can have a dedicated section for getting the token and then refer to it from all other sections,  that might work, will try tomorrow

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+
+image::mcp_server_oidc_devui.png[MCP Server and OIDC in DevUI,align="center"]
+
+Select an OpenId Connect card and https://quarkus.io/guides/security-openid-connect-dev-services#develop-service-applications[login to Keycloak] using an `alice` name and an `alice` password.
```
```suggestion
Select the OpenId Connect card and https://quarkus.io/guides/security-openid-connect-dev-services#develop-service-applications[login to Keycloak] using an `alice` name and an `alice` password.
```

I'm not sure _select_ is the right word

sberyozkin replied:
Sure, I'll just use `choose`

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+[NOTE]
+====
+You can login to other providers such as `Auth0` or https://quarkus.io/guides/security-openid-connect-providers#github[GitHub] from OIDC DevUI as well. The only requirement is to update your application registration to allow callbacks to DevUI. For example, see how you can https://quarkus.io/guides/security-oidc-auth0-tutorial#looking-at-auth0-tokens-in-the-oidc-dev-ui[login to Auth0 from Dev UI] ...
+
```
```suggestion
===
```

otherwise, you are still in the admonition.

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+Use the `Call` action to make a tool call and get a response which contains the `alice` user name:
+
+image::mcp_server_tool_response.png[MCP Server tool response,align="center"]
+
```
It's great, but I'm wondering if a short screen recording would not be better (without voice over).

sberyozkin replied:
Hi @cescoffier I propose to do it in follow up presentations, or may be in Insights, I'll have to refresh my knowledge again how to do recordings on Fedora. I did it before, but the software is not on the bew laptop.

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+image::mcp_server_tool_response.png[MCP Server tool response,align="center"]
+
+[[mcp-inspector]]
+=== Use MCP Inspector
```
== Step 3 - Invoking an MCP tool requiring authentication with the MCP Inspector

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+
+[[mcp-inspector]]
+=== Use MCP Inspector
+
```
```suggestion
The [MCP Inspector](https://github.com/modelcontextprotocol/inspector) is an interactive developer utility for testing and debugging MCP servers. It is handy to understand what's going on. Let's use this inspector to invoke our MCP server with authentication.
```

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+npx @modelcontextprotocol/inspector
+----
+
+Next, get and copy the access token as explained in the <<mcp-server-devui>> section above. You can use the same access token that you may have already used when working through that section.
```
Here I would do things slightly differently. I would run the application in "prod" mode - no dev UI, _real_ keycloak. Do you think it would be possible?

sberyozkin replied:
@cescoffier I may have missed it, but yeah, now we do it with GitHub

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+
+image::mcp_inspector_tool_call.png[MCP Inspector Tool Call,align="center"]
+
+=== Use curl
```
```suggestion
== == Step 4 - Invoking an MCP tool requiring authentication with curl
```

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+=== Use curl
+
+Finally, let's use `curl` and also learn a little bit how the MCP protocol and MCP SSE transport work. 
+
```
What about adding something like this (and a picture):

When invoking an MCP tool using the SSE transport, there are at least two requests:

1. A first request connects to the server using SSE - this _back channel_ receives the invocation results and notifications
2. A second request (`POST`) invokes the tool. The invocation result is transmitted on the SSE _back channel_.

sberyozkin replied:
@cescoffier Sorry, missed it too, good idea, if you still think it is worth adding it, please create a diagram with your nice tools, may be on Monday ? and We  can add it to the `curl` section, I can try to sketch it as well, but you have a very clear picture how it should be presented, so yeah, please do it and we can add it tomorrow

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+
+== Create MCP SSE server
+
+First, let's create a secure Quarkus MCP SSE server that requires authentication during the initial SSE handshake and the tool access.
```
```suggestion
First, let's create a secure Quarkus MCP SSE server that requires authentication during the initial Server-Sent Events (SSE) handshake and the tool access.
```

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+
+Finally, let's use `curl` and also learn a little bit how the MCP protocol and MCP SSE transport work. 
+
+First, access the main SSE endpoint without the access token:
```
Same if possible I would use prod mode. I think it's important to show that this feature is not only a _dev_ thing, it's a real deal.

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+
+[source,shell]
+----
+curl -v -H "Authorization: Bearer ey..." -H "Content-Type: application/json" --data @call.json http://localhost:8080/mcp/messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
```
I know it will automatically switch to POST as you have a body, but would it be possible to specify the verb? Not everyone is a curl expert :-)

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+}
+----
+
+Now look in the first curl window and observe a response which contains the `alice` user name:
```
```suggestion
Now look the SSE connection and observe a response which contains the `alice` user name:
```

cescoffier commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+
+== Conclusion
+
+In this blog post, we have explained how you can easily create a secure Quarkus MCP SSE server, obtain a bearer access token and test it in 
```
```suggestion
In this blog post, we have explained how you can easily create an authenticated Quarkus MCP SSE server, obtain a bearer access token and test it in 
```

cescoffier replied:
in what? Something missing :-)

sberyozkin replied:
@cescoffier Hi, lol, I probably jumped to some other section, sure, will fix

sberyozkin commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+
+While it will take a bit of time for the MCP authorization part of the MCP specification be finalized and widely supported, one thing you can be certain about is that MCP authorization compliant clients will be able to login users with the OAuth2 authorization code flow and use bearer tokens to access MCP servers on their behalf.
+
+In the meantime, you can use any MCP client that can accept a bearer access token and send it to the MCP server to start experimenting with MCP authorization.
```
@cescoffier 

>This flow is quite similar to what is defined in the latest MCP specification (add link)

Unfortunately it is not. The latest specification describes an authorization code flow, just having some tool like MCP inspector be able to pass the token is not related to the specification as such

sberyozkin replied:
@cescoffier But that said, I can try to clarify that passing the bearer token is part of the latest MCP authprization proposal, thanks

sberyozkin replied:
@cescoffier I'll add a link to https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization#2-6-access-token-usage

sberyozkin commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+
+https://modelcontextprotocol.io/specification/2025-03-26[The latest version of the Model Context Protocol (MCP) specification] introduces an https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization[authorization] flow.
+
+While it will take a bit of time for the MCP authorization part of the MCP specification be finalized and widely supported, one thing you can be certain about is that MCP authorization compliant clients will be able to login users with the OAuth2 authorization code flow and use bearer tokens to access MCP servers on their behalf.
```
@cescoffier May be I'm missing something,

>you can already add authentication to client and server following the previous MCP version (Add version date).

Where exactly in the previous MCP version, at https://modelcontextprotocol.io/specification/2024-11-05, one can do the authentication ?

sberyozkin replied:
Here I'm trying to prepare users for what is coming with the `latest` spec, i.e, no matter how the MCP client acquires the token (and it must be authorization code flow according to the latest spec, with client_credentials mentioned but omitted from the Microsoft PR), MCP server will receive a bearer token

maxandersen replied:
the old mcp client protocol is http based which support setting these headers - hence it works with existing one.

cescoffier replied:
@sberyozkin can you apply my suggestion?

sberyozkin replied:
Sounds good, @cescoffier, @maxandersen, indeed, it was already HTTP based before.

How does it look now ?

sberyozkin commented:
@cescoffier Thanks, sure, I'll have a look at having it run with jbang, I'll apply all the suggestion but without talking the versions and then let's try to do one more review round

maxandersen commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+
+== Introduction
+
+https://modelcontextprotocol.io/specification/2025-03-26[The latest version of the Model Context Protocol (MCP) specification] introduces an https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization[authorization] flow.
```
```suggestion
```

I would just leave this out or put link somewhere below.

cescoffier commented:
I've another long train journey tomorrow. I will do another review tomorrow.

maxandersen commented on _posts/2025-04-25-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,399 @@
+
+In this post, we will create a https://github.com/quarkiverse/quarkus-mcp-server[Quarkus MCP SSE Server] that requires authentication to access its tools.
+
+We will use Keycloak to login and use a Keycloak JWT access token to access the server with `Quarkus MCP SSE Server Dev UI` in the dev mode.
```
```suggestion
We will first use Keycloak as an OpenID provider to login and use a Keycloak JWT access token to access the server with `Quarkus MCP SSE Server Dev UI` in the dev mode.
```

maxandersen commented on _posts/2025-04-25-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,399 @@
+
+We will use Keycloak to login and use a Keycloak JWT access token to access the server with `Quarkus MCP SSE Server Dev UI` in the dev mode.
+
+We will use GitHub to login and use a GitHub binary access token to access the server in the prod mode with both https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector] and the `curl` tools.
```
```suggestion
Secondly we will show how to use GitHub OpenID to login and use a GitHub binary access token to access the server in the prod mode with both https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector] and the `curl` tools.
```

maxandersen replied:
not sure if OpenID is the correct term but idea is to just highlight its two different ways, keycloak not the only option and openid/oauth is that common enabler.

sberyozkin replied:
Sure, @maxandersen, I'll tune it a tiny bit and will apply

sberyozkin replied:
OpenID is fine but is more general SSO term, OpenId Connect, being built on top of OAuth2, is an OpenID in action

sberyozkin replied:
GitHub and many other social providers are pure OAuth2, they don't offer ID tokens

maxandersen commented on _posts/2025-04-25-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,399 @@
+
+The OIDC configuration is provided in the dev mode automatically by https://quarkus.io/guides/security-openid-connect-dev-services[Dev Services for Keycloak]. It creates a default realm, client and adds two users, `alice` and `bob`, for you to get started with OIDC immediately. You can also register a custom Keycloak realm to work with the existing realm, client and user registrations.
+
+No problems if you do not work with Keycloak, see the <<mcp-server-devui>> section for more details.
```
i'm not folowing what that sentence is trying to say? If we are running in devmode why is "working with keycloak"  a problem?

sberyozkin replied:
@maxandersen I'm trying to say that Auth0 users, GitHub users can also use this DevUI and login to their providers from Dev UI, but I'll try to rephase it

sberyozkin replied:
@maxandersen 
The reason I don't show GitHub in devmode is that for that to work, users would have to allow a redirect to OIDC Dev UI, something like `http://localhost:8080/q/dev-ui/io.quarkus.quarkus-oidc/github-provider` (or similar), see https://quarkus.io/guides/security-openid-connect-dev-services#dev-ui-all-oidc-providers (it shows Google) but also in our Auth0 guide (the link is in the section below here).

This can be done, but with GitHub, they only allow a single redirect URL, but we also need to support a Login endpoint redirect in prod, so it can become messy, users having to update their GitHub OAuth2 registrations dependeing on whether it is devmode or prod....

sberyozkin replied:
So yes, in Dev UI, we have all out of the box already, and showing Keycloak is simpler for the initial check

maxandersen commented on _posts/2025-04-25-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,399 @@
+Now, please stop the server: we are going to prepare it to run in the prod mode next.
+
+== Step 3 - Access MCP server in the prod mode
+
```
Maybe add something like:

"Before it was all in dev mode - using Quarkus devservices to try things out. Now lets move to prod mode. If you have an keycloak instance running you can use that (see link); but to illustrate how oauth works with more than just keycloak and something many of you have direct access we will show prod mode using GitHub OpenID."

?

maxandersen commented on _posts/2025-04-25-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,399 @@
+== Conclusion
+
+In this blog post, we have explained how you can easily create a secure Quarkus MCP SSE server, obtain an access token and use it to access the MCP server tool in the dev mode with `Quarkus MCP SSE Server Dev UI` and the prod mode with both the https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector] and the curl tools.
+
```
maybe add:

"Notice that there is no real difference from adding oauth to MCP or how it is done for a REST endpoint. The most complex part is to get the settings configured for your OpenID provider - but when all is done you just apply a few annotations to mark relevant methods as secure and Quarkus handle the authentication for you."

sberyozkin replied:
I'll try

maxandersen reviewed (dismissed):
a few minor nitpicks for clarity but otherwise really good stuff!

cescoffier commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,403 @@
+
+While it will take a bit of time for the new MCP specification to be widely supported, you can already add authentication to client and server following the https://modelcontextprotocol.io/specification/2024-11-05[previous MCP version].
+
+Indeed, you can use any MCP client that can receive an access token and pass it to the MCP server.
```
```suggestion
You only need an MCP client that can receive an access token and pass it to the MCP server and, obviously, an MCP server that verifies the token.
```

cescoffier commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,403 @@
+
+The SSE connection is created.
+
+Now open another window and use the same access token to initialize the curl as MCP client, and access the tool, using the value of the `data` property to build the target URL.
```
Add something like:

We cannot invoke the tool directly, we first need to follow the MCP handshake protocol:

cescoffier commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,403 @@
+
+== Conclusion
+
+In this blog post, we have explained how you can easily create a secure Quarkus MCP SSE server, obtain an access token and use it to access the MCP server tool in dev mode with `Quarkus MCP SSE Server Dev UI` and prod mode with both the https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector] and the curl tools.
```
```suggestion
In this blog post, we explained how you can easily create a authenticated Quarkus MCP SSE server, obtain an access token and use it to access the MCP server tool in dev mode with `Quarkus MCP SSE Server Dev UI` and prod mode with both the https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector] and the curl tools.
You can use any MCP client that allows passing a bearer token to the server.
```

cescoffier commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,403 @@
+
+Notice, that there is no real difference in how OAuth2 is done for either Quarkus MCP server or REST endpoints. The most complex part is to get the settings configured correctly for your OAuth2 provider - but when all is done you just apply a few annotations to mark relevant methods as secure and Quarkus handles the authentication for you.
+
+The Quarkus team is keeping a close eye on the MCP Authorization specification evolution and working on having all possible MCP Authorization scenarios supported.
```
```suggestion
This blog post uses the previous version of the MCP protocol. The Quarkus team is keeping a close eye on the new MCP Authorization specification evolution and working on having all possible MCP Authorization scenarios supported.
```

cescoffier commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,403 @@
+
+The Quarkus team is keeping a close eye on the MCP Authorization specification evolution and working on having all possible MCP Authorization scenarios supported.
+
+Stay tuned for more updates !
```
```suggestion
Stay tuned for more updates!
```

cescoffier reviewed (changes requested):
Wow! Sergei, this is great, it's going to be awesome!
I've made a few suggestions. There is a flaw in the flow when we configure the MCP server with GitHub.

sberyozkin commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,403 @@
+
+Indeed, you can use any MCP client that can receive an access token and pass it to the MCP server.
+
+In this post, we will create a https://github.com/quarkiverse/quarkus-mcp-server[Quarkus MCP SSE Server] that requires authentication to access its tools.
```
Just a tiny typo update

sberyozkin commented on _posts/2025-04-24-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,227 @@
+<1> Provide a tool that can return a name of the current user. Note the `user-name-provider` tool name, you will use it later for a tool call.
+<2> Require authenticated tool access. See also how the main MCP SSE endpoint is secured in the <<configuration>> section below.
+<3> Use the injected `SecurityIdentity` to return the current user's name.
+
```
@cescoffier Hi, I've updated the second point which will now read: `Require authenticated tool access - yes, the only difference with an unauthenticated MCP server tool is `@Authenticated`, that's it!.` - thanks; and indeed, the sentence that follows there refers to also authenticating the main sse endpoint.
Indeed, we'd need to make that one public to support fixing authenticated and unauthenticated tool calls

sberyozkin replied:
I.e, just `@Athenticated` alone protects the tool only

sberyozkin commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,403 @@
+    AccessTokenCredential accessToken; <2>
+
+    @Inject
+    Template accessTokenPage;
```
@cescoffier Yeah, I link the source at the start of the blog, and it is just a basic Qute template, I thought its content was not really important as we already have quote a few images, but let me link to it in the demo source

sberyozkin commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,403 @@
+%prod.quarkus.oidc.provider=github <1>
+%prod.quarkus.oidc.application-type=service <2>
+
+%prod.quarkus.oidc.login.provider=github <3>
```
@cescoffier Sure, yeah, I've reworked it a bit to try to avoid some upfront references to the config

sberyozkin commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,403 @@
+}
+----
+
+and send the initialization notification:
```
@cescoffier This was done in the previous request, I've tried the text for both requests a little bit

sberyozkin commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,403 @@
+
+=== Install and run application
+
+First, add `quarkus.package.jar.type=uber-jar` to the application.properties.
```
Hi @cescoffier 

Sure, I believe we use `jbang` to make it easy to plugin MCP servers into Claude and other MCP clients, so, once Claude will start supporting logging in users, this demo will just work there with `jbang`.

But I've added a note:

```
[NOTE]
====
The `quarkus.package.jar.type=uber-jar` property is set to make it possible to run the MCP server with `jbang`.

If you prefer, you can simply package and run the MCP server instead:

[source,shell]
----
mvn package
java -jar target/quarkus-app/quarkus-run.jar
----
====
```

Hope it covers both options

cescoffier replied:
Claude only supports stdio transport (for now).
So, switch to what Quarkus recommends, not planning on something we don't know.

sberyozkin commented:
@cescoffier Thanks Clement for a very thorough review, very appreciated. 

There have been quite a few comments, I hope I've addressed all of them, either applying the suggestions directly or trying to react to some suggestion with an update, where I felt the direct suggestion application may not work best.

I'm going to check for various typos, the only 2 main things that I believe are open are:
* Short video, as I commented, I propose to do it in a more dedicated session about MCP security
* In the curl section, adding a nice little picture to help users visualize SSE flow would be great, but that suggestion was added 3 days ago and I missed it, and I reworked the curl section since then. So if you would like please propose the one tomorrow on Monday, or we can just keep it as is.

Have another quick look tomorrow, I think the PR is very close to being merged now, cheers

sberyozkin commented:
I published it locally and did a few minor follow up updates like restoring a line break between `Firstly` and `Secondly` in the intro as it looks like one a big sentence. Fixed a few broken links, overall is not too bad, it is still an intro :-), we can add many more details and content later as well :-)

cescoffier commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,431 @@
+[[initial-mcp-server]]
+== Step 1 - Create an MCP server using the SSE transport
+
+First, let's create a secure Quarkus MCP SSE server that requires authentication to establish Server-sent Events (SSE) connection and also when invoking the tools.
```
```suggestion
First, let's create a secure Quarkus MCP SSE server that requires authentication to establish Server-Sent Events (SSE) connection and also when invoking the tools.
```

cescoffier commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,431 @@
+image::mcp_server_tool_response.png[MCP Server tool response,align="center"]
+
+All is good in dev mode; it is time to see how it will work in prod mode.
+Before that, make sure you stop the MCP server whuch runs in dev mode.
```
```suggestion
Before that, stop the MCP server, which runs in dev mode.
```

cescoffier commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,431 @@
+
+[NOTE]
+====
+Note the use of the `%prod.` prefixes. It ensures the configuration properties prefixed with `%prod.` are only effective in prod mode and not interfering with dev mode. 
```
```suggestion
Note the use of the `%prod.` prefixes. It ensures the configuration properties prefixed with `%prod.` are only effective in prod mode and do not interfere with dev mode. 
```

cescoffier commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,431 @@
+
+=== Implement Login endpoint
+
+Currently, MCP clients can not use the authorization code flow themselves, therefore we implement an OAuth2 login endpoint which will return a GitHub token for the user to use it with MCP clients which can work with bearer tokens.
```
```suggestion
Currently, MCP clients can not use the authorization code flow themselves. Therefore, we implement an OAuth2 login endpoint that will return a GitHub token for the user to use with MCP clients, which can work with bearer tokens.
```

cescoffier commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,431 @@
+data: /messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
+----
+
+The SSE connection is created. 
```
```suggestion
The SSE connection is created.  Note the unique path in the received `data`, we need this path to invoke the tools.
```

cescoffier reviewed (changes requested):
How to start the server in prod mode needs to be reworked a bit. 
The rest is just a few typos.

sberyozkin commented:
@cescoffier 

>So, switch to what Quarkus recommends, not planning on something we don't know.

The fact that Claude does not support SSE is not really important, it is just that this demo would be ready to be plugged in We just show how to start with jbang - we don't recommend using uber jar - I added a note to tell users that the only reason we do it is to support a jbang start up

sberyozkin commented:
Anyway, I'll just update as you proposed, but add a note about jbang instead

maxandersen commented:
> @cescoffier
> 
> > So, switch to what Quarkus recommends, not planning on something we don't know.
> 
> The fact that Claude does not support SSE is not really important, it is just that this demo would be ready to be plugged in We just show how to start with jbang - we don't recommend using uber jar - I added a note to tell users that the only reason we do it is to support a jbang start up

the uber jar is not required for jbang.  jbang supports non-uber-jar. its java that is the limit here. uber-jar is just so its easier to distributed and faster to start

cescoffier commented:
Faster to start? I really doubt. 
Easier to start, maybe.

maxandersen commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,429 @@
+
+[NOTE]
+====
+You can also run the MCP server with `jbang`.
```
```suggestion
You can also run the MCP server from its Maven coordinates directly with `jbang`.
```

maxandersen commented on _posts/2025-04-28-secure-mcp-sse-server.adoc:
```diff
@@ -0,0 +1,429 @@
+====
+You can also run the MCP server with `jbang`.
+
+You have to set the `quarkus.package.jar.type=uber-jar` property first, and then install and run it with `jbang`:
```
```suggestion
When using Maven coordinates its advantageous to use `quarkus.package.jar.type=uber-jar` property first, and then install and run it with `jbang`:
```

that setting is not needed for jbang.

you can also just run the generated .jar directly with java or jbang.  so adjusted to the correct thing above.

maxandersen replied:
actually - you sholdn't even needed to use uber-jar since you are not using @fatjar thus dependencies will be resolved anyway. So its unnecessary - its mvn install that is needed.

sberyozkin replied:
@maxandersen Indeed, thanks for the hint, I just dropped a ref to the uber jar property for now.

sberyozkin commented:
Thanks @maxandersen, @cescoffier, it should be ready to go now ?
